ml-fundamentals/ml-experimentation.md

```diff
@@ -0,0 +1,147 @@
+# Model Experimentation
+
+Machine learning model experimentation involves uncertainty around the expected model results and future operationalization.
+To handle this uncertainty as much as possible, we propose a semi-structured process, balancing between engineering/research best practices and rapid model/data exploration.
+
+## Model experimentation goals
+
+1. **Performance**: Find the best performing solution
+2. **Operationalization**: Keep an eye towards production, making sure the operationalization is feasible
+3. **Code quality** Maintain code and artifacts quality
+4. **Reproducibility**: Keep research active by allowing experiment tracking and reproducibility
+5. **Collaboration**: Foster the collaboration and joint work of multiple people on the team
+
+## Model experimentation challenges
+
+- **Trial and error process**: Difficult to plan and estimate durations and capacity.
+- **Quick and dirty**: We want to fail fast and get a sense of what’s working efficiently
+- **Collaboration**: How do we form a team-wide trial and error process and effective brainstorming
+- **Code quality**: How do we maintain the quality of non-production code during research
+- **Operationalization**: Switching between approaches might have a significant impact on operationalization (e.g. GPU/CPU, batch/online, parallel/sequential, runtime environments)
+
+Creating an experimentation framework which facilitates rapid **experimentation**, **collaboration**,
+experiment and model **reproducibility**, **evaluation**  and **defined APIs**,
+and lets each team member focus on the model development and improvement,
+while trusting the framework to do the rest.
+
+The following tools and guidelines are aimed at achieving experimentation goals as well as addressing the aforementioned challenges.
+
+## Tools and guidelines for successful model experimentation
+
+- [Virtual environments](#virtual-environments)
+- [Source control and folder/package structure](#source-control-and-folder-package-structure)
+- [Experiment tracking](#experiment-tracking)
+- [Datasets and models abstractions](#datasets-and-models-abstractions)
+- [Model evaluation](#model-evaluation)
+
+### Virtual environments
```

Comment:
We should mention that virtual environments allow us to be consistent across our local dev envs as well as with compute resources. We should use the same virt evn "source" *.txt/*yml. Anywhere we can build from the same source allows us to remove errors due to inconsistency.